AX: Make AXTextRuns a shared_ptr in the AXPropertyValueVariant

hoffmanjoshua opened the pull request:
#### 6b807dc25d6124423dcc1d436bc731f374183704
<pre>
AX: Make AXTextRuns a shared_ptr in the AXPropertyValueVariant
<a href="https://bugs.webkit.org/show_bug.cgi?id=292575">https://bugs.webkit.org/show_bug.cgi?id=292575</a>
<a href="https://rdar.apple.com/problem/150721813">rdar://problem/150721813</a>

Reviewed by Tyler Wilcock.

The size of a variant is relative to the largest type within it, and in the
AXPropertyValueVariant, that was the AXTextRuns, which was 32 bytes (in part
due to memory alignment). This made the overall variant 48 bytes after
padding.

To reduce this size, we can instead store shared_ptr pointing to the text runs,
which only uses 16 bytes. This makes the overall variant 32 bytes after
padding.

While making this change, I also moved the type we used to store the marked
text marker range to be a shared_ptr, since this was the next-largest type at
24 bytes.

This saves 33% of memory per-property that we store in the properties vector.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::textRuns const):
(WebCore::AXIsolatedObject::textInputMarkedTextMarkerRange const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/294565@main">https://commits.webkit.org/294565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c53c08dc416e780fd70c98750ce94439fd2fe976

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102250 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21918 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12233 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107409 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52886 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104289 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22226 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30425 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77802 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34789 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105256 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17200 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92300 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58140 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17031 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10328 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52244 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86860 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10398 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109785 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29382 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21657 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86784 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29746 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88503 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86372 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21982 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31179 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8895 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23624 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29310 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34605 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29121 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32444 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30680 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
```diff
@@ -293,8 +293,10 @@ WTF::TextStream& operator<<(WTF::TextStream&, AXProperty);
 
 using AXPropertySet = HashSet<AXProperty, IntHash<AXProperty>, WTF::StrongEnumHashTraits<AXProperty>>;
 
+using AXIsolatedCharacterRange = std::pair<Markable<AXID>, CharacterRange>;
```
`AXIsolated` doesn't feel right here...there's nothing specific to isolated objects about this type. Maybe `AXIDAndCharacterRange`? Or maybe not re-name this type? I don't like typedefs personally because it makes it harder to know what you're working with, so it has to be greatly shorter than the type to be worth it IMO, e.g.:

`using AXPropertySet = HashSet<AXProperty, IntHash<AXProperty>, WTF::StrongEnumHashTraits<AXProperty>>;`

hoffmanjoshua replied:
I chose to rename this type since there are places we have to use it twice in the same line, which would end up looking like this:
```cpp
setProperty(AXProperty::TextInputMarkedTextMarkerRange, std::make_shared<std::pair<Markable<AXID>, CharacterRange>>(std::pair<Markable<AXID>, CharacterRange>(range.start().objectID(), *characterRange)));
```

I like the name you suggested. I can switch to that for now and see if something more concise comes to mind!

hoffmanjoshua committed:
AX: Make AXTextRuns a shared_ptr in the AXPropertyValueVariant
https://bugs.webkit.org/show_bug.cgi?id=292575
rdar://problem/150721813

Reviewed by Tyler Wilcock.

The size of a variant is relative to the largest type within it, and in the
AXPropertyValueVariant, that was the AXTextRuns, which was 32 bytes (in part
due to memory alignment). This made the overall variant 48 bytes after
padding.

To reduce this size, we can instead store shared_ptr pointing to the text runs,
which only uses 16 bytes. This makes the overall variant 32 bytes after
padding.

While making this change, I also moved the type we used to store the marked
text marker range to be a shared_ptr, since this was the next-largest type at
24 bytes.

This saves 33% of memory per-property that we store in the properties vector.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::textRuns const):
(WebCore::AXIsolatedObject::textInputMarkedTextMarkerRange const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/294565@main
